File: modules/studybolt-on-device-ai/android/src/main/java/com/studybolt/ondeviceai/StudyBoltOnDeviceAIModule.kt
package com.studybolt.ondeviceai

import android.os.Build
import com.google.mlkit.genai.common.DownloadStatus
import com.google.mlkit.genai.common.FeatureStatus
import com.google.mlkit.genai.prompt.Generation
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class StudyBoltOnDeviceAIModule : Module() {
  private val generativeModel by lazy { Generation.getClient() }

  override fun definition() = ModuleDefinition {
    Name("StudyBoltOnDeviceAI")

    AsyncFunction("getAvailability") Coroutine {
      currentAvailability()
    }

    AsyncFunction("downloadModel") Coroutine {
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
        return@Coroutine unavailable("On-device tutoring requires Android 8 or later.")
      }

      if (generativeModel.checkStatus() == FeatureStatus.DOWNLOADABLE) {
        generativeModel.download().collect { status ->
          if (status is DownloadStatus.DownloadFailed) {
            throw OnDeviceAIException("Gemini Nano could not be downloaded.", status.e)
          }
        }
      }
      currentAvailability()
    }

    AsyncFunction("generate") Coroutine { prompt: String, instructions: String ->
      if (generativeModel.checkStatus() != FeatureStatus.AVAILABLE) {
        throw OnDeviceAIException("Gemini Nano is not ready on this phone.")
      }

      try {
        val boundedPrompt = buildString {
          append(instructions.take(2_000))
          append("\n\n")
          append(prompt.take(10_000))
        }
        val response = generativeModel.generateContent(boundedPrompt)
        val content = response.candidates.firstOrNull()?.text?.trim().orEmpty()
        if (content.isEmpty()) throw OnDeviceAIException("Gemini Nano returned an empty response.")
        mapOf(
          "text" to content.take(4_000),
          "provider" to "gemini-nano"
        )
      } catch (error: OnDeviceAIException) {
        throw error
      } catch (error: Throwable) {
        throw OnDeviceAIException("Gemini Nano could not answer that right now.", error)
      }
    }

    OnDestroy {
      generativeModel.close()
    }
  }

  private suspend fun currentAvailability(): Map<String, String> {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
      return unavailable("On-device tutoring requires Android 8 or later.")
    }
    return when (generativeModel.checkStatus()) {
      FeatureStatus.AVAILABLE -> available("available", "Gemini Nano is ready on this device.")
      FeatureStatus.DOWNLOADABLE -> available("downloadable", "Gemini Nano can be downloaded for private, on-device tutoring.")
      FeatureStatus.DOWNLOADING -> available("downloading", "Gemini Nano is downloading on this device.")
      else -> unavailable("This Android phone does not currently support Gemini Nano Prompt API.")
    }
  }

  private fun available(status: String, reason: String) = mapOf(
    "status" to status,
    "provider" to "gemini-nano",
    "reason" to reason
  )

  private fun unavailable(reason: String) = available("unavailable", reason)
}

private class OnDeviceAIException(
  message: String,
  cause: Throwable? = null
) : CodedException(message, cause)

File: modules/studybolt-on-device-ai/ios/StudyBoltOnDeviceAIModule.swift
import ExpoModulesCore
import Foundation

#if canImport(FoundationModels)
import FoundationModels
#endif

public final class StudyBoltOnDeviceAIModule: Module {
  public func definition() -> ModuleDefinition {
    Name("StudyBoltOnDeviceAI")

    AsyncFunction("getAvailability") {
      return currentAvailability()
    }

    AsyncFunction("downloadModel") {
      // Apple manages the Foundation Model download at the system level.
      return currentAvailability()
    }

    AsyncFunction("generate") { (prompt: String, instructions: String, promise: Promise) in
      #if canImport(FoundationModels)
      if #available(iOS 26.0, *) {
        let model = SystemLanguageModel.default
        guard model.isAvailable else {
          promise.reject(OnDeviceAIUnavailableException())
          return
        }

        let boundedInstructions = String(instructions.prefix(2_000))
        let boundedPrompt = String(prompt.prefix(10_000))
        Task {
          do {
            let session = LanguageModelSession(model: model) {
              boundedInstructions
            }
            let response = try await session.respond(to: boundedPrompt)
            let content = response.content.trimmingCharacters(in: .whitespacesAndNewlines)

            guard !content.isEmpty else {
              promise.reject(EmptyOnDeviceAIResponseException())
              return
            }

            promise.resolve([
              "text": String(content.prefix(4_000)),
              "provider": "apple-intelligence"
            ])
          } catch {
            promise.reject(error)
          }
        }
        return
      }
      #endif

      promise.reject(OnDeviceAIUnavailableException())
    }
  }

  private func currentAvailability() -> [String: String] {
    #if canImport(FoundationModels)
    if #available(iOS 26.0, *) {
      switch SystemLanguageModel.default.availability {
      case .available:
        return availability("available", reason: "Apple Intelligence is ready on this device.")
      case .unavailable(.appleIntelligenceNotEnabled):
        return availability("unavailable", reason: "Turn on Apple Intelligence in Settings to use on-device tutoring.")
      case .unavailable(.deviceNotEligible):
        return availability("unavailable", reason: "This iPhone does not support Apple Intelligence.")
      case .unavailable(.modelNotReady):
        return availability("downloading", reason: "Apple Intelligence is preparing its on-device model.")
      @unknown default:
        return availability("unavailable", reason: "Apple Intelligence is not available right now.")
      }
    }
    #endif

    return availability("unavailable", reason: "On-device tutoring requires iOS 26 or later and a supported iPhone.")
  }

  private func availability(_ status: String, reason: String) -> [String: String] {
    [
      "status": status,
      "provider": "apple-intelligence",
      "reason": reason
    ]
  }
}

internal final class OnDeviceAIUnavailableException: Exception, @unchecked Sendable {
  override var reason: String {
    "The phone's on-device AI is not available."
  }
}

internal final class EmptyOnDeviceAIResponseException: Exception, @unchecked Sendable {
  override var reason: String {
    "The phone's on-device AI returned an empty response."
  }
}
